### PR TITLE
Time: Improve Generic Time handling

### PR DESCRIPTION
**Pull request by nilnilnil**
@MrChrisW @moollaza @bsstoner 

@bsstoner I could probably use some JS cleanup :)

-----
IA Page: https://duck.co/ia/view/time

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata post coming soon

File: share/spice/time/time.js [hunk 1]
```diff
@@ -6,21 +6,36 @@
...
 
+
         var script = $('[src*="/js/spice/time/"]')[0],
             source = $(script).attr("src"),
             // Query is normalized as we'll normalize the generated strings.
             // if we have a comma separated query it is in the form:
             // "town, state, country"  but state is optional
             query = decodeURIComponent(source.match(/time\/([^\/]+)/)[1]).toLowerCase().split(','),
+            isGeneric = /\/generic\//.test(source),
+            callParameters = decodeURIComponent(source).split('/'),
```
Comment (bsstoner) [inline share/spice/time/time.js]: Maybe use `decodeURIComponentSafe(source)` here to avoid page crashing errors?

File: share/spice/time/time.js [hunk 2]
```diff
@@ -50,7 +65,11 @@
             day: dateObj.getDate(),
             monthName: months[dateObj.getMonth()],
             year: dateObj.getFullYear(),
-            placeName: chosen.geo.state ? (chosen.geo.name + ", " + chosen.geo.state) : chosen.geo.name,
+            placeName: chosen.overridePlaceName ?
+                chosen.overridePlaceName :
+                chosen.geo.state ?
+                  (chosen.geo.name + ", " + chosen.geo.state) :
+                  chosen.geo.name,
```
Comment (bsstoner) [inline share/spice/time/time.js]: This ternary is getting a little out of control, maybe assign to a variable above?

File: share/spice/time/time.js [hunk 3]
```diff
@@ -6,21 +6,36 @@
...
             chosen;
 
-        for(var i = 0; i < api.locations.length; i++) {
-            // query[0] = state
-            // query[len-1] = country
-            if(DDG.stringsRelevant(query[0], api.locations[i].geo.name) &&
-                    DDG.stringsRelevant(query[query.length-1], api.locations[i].geo.country.name)) {
-                chosen = api.locations[i];
-                break;
+        if (isGeneric && callParameters.length === 7) {
```
Comment (bsstoner) [inline share/spice/time/time.js]: why 7? maybe leave a comment?

File: share/spice/time/time.js (hunk 2 above)
Comment (nilnilnil) [inline share/spice/time/time.js]: I knew I'd get shit for this. OK!

**Comment by nilnilnil** [comment]
@bsstoner all done.

**Comment by nilnilnil** [comment]
@bsstoner Thanks for the feedback. Handing off to original maintainer @MrChrisW for final review.

I think there is a lot more low hanging fruit here. This still doesn't solve the query `time in new york`.

**Comment by MrChrisW** [comment]
@nilnilnil Have tested, looks good to me.

Any reason you also check the length of `callParameters`? Assuming 'generic' is found in the call all parameters "should" be found, doesn't make much difference... anyway a comment might be good like @bsstoner suggested. 

I'll hand this to @moollaza to merge :+1: 

:christmas_tree:

**Comment by nilnilnil** [comment]
I added a comment. Is it confusing? I'm happy to revise. I check it because I don't want to assume everything will be defined properly. It's probably a safe bet though. Didn't see a reason to risk it.

**Comment by MrChrisW** [comment]
@nilnilnil Nah it's fine! Makes sense to check the array items exist before using them.  :+1:  Merry Christmas

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=time in Melbourne)
                
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/time_mobile.png)

**Comment by jagtalon** [comment]
:+1: 

giphy

**Pull request merged**